File: Data-structure/segment_tree.py
class SegmentTree(object):
	def __init__(self, data):
		self.data = data
		self.tree = [0]*len(data)*4
		self.treeInit(1, 1, 8)

	def treeInit(self, node, left, right):
		if left == right:
			self.tree[node] = self.data[left]
			return 
		_left = node*2
		_right = node*2+1
		mid = (int)(left+right)/2
		self.treeInit(_left, left, mid)
		self.treeInit(_right, mid+1, right)
		self.tree[node] = self.tree[_left]+self.tree[_right]

	def query(self, node, left, right, i, j):
		if i > right or j < left:
			return 0
		elif left >= i and right <= j:
			return self.tree[node]
		_left = node*2
		_right = node*2+1
		mid = (int)(left+right)/2
		p1 = self.query(_left, left, mid, i, j)
		p2 = self.query(_right, mid+1, right, i, j)
		return p1+p2

	def update(self, node, left, right, i, j, val):
		if i > right or j < left:
			return 
		elif left >= i and right <= j:
			self.tree[node] = val
			return 
		_left = node*2
		_right = node*2+1
		mid = (int)(left+right)/2
		self.update(_left, left, mid, i, j, val)
		self.update(_right, mid+1, right, i, j, val)
		self.tree[node] = self.tree[_left]+self.tree[_right]


if __name__ == '__main__':
	s = SegmentTree([0,1,2,3,4,5,6,7,8])
	s.update(1,1,8,1,1,10)
	print(s.query(1,1,8,1,3))

File: Data-structure/segment_tree.cpp
# include <iostream>
# include <cstdio>
using namespace std;
#define mx 100001

int arr[mx], tree[mx*4];

void prepareTree(int node, int l, int r) {
	if(l == r) {
		tree[node] = arr[l];
		return;
	}
	int left = (node * 2);
	int right = (node * 2 + 1);
	int mid = (int) (l + r) / 2;
	prepareTree(left, l, mid);
	prepareTree(right, mid +1 , r);
	tree[node] = tree[left] + tree[right];
}

int query(int node, int l, int r, int i, int j) {
	if(i > r || j < l) return 0; // outside the range
	else if(i <= l && j>=r) return tree[node]; // inside the range
	int left = (node * 2);
	int right = (node*2 +1);
	int mid = (l + r) / 2;
	int p1 = query(left, l, mid, i, j);
	int p2 = query(right, mid+1, r, i, j);
	return p1 + p2;

}

void update(int node , int l, int r, int i, int j, int val) {
	if(i > r || j < l) return;

	else if (i <= l && j >= r) {
		tree[node] = val;
		return;
	}

	int left = (node * 2);
	int right = (node * 2 + 1);
	int mid = (l + r) / 2;

	update(left, l, mid, i, j, val);
	update(right, mid+1, r, i, j, val);
	tree[node] = tree[left] + tree[right];
}

int main() {
	int n = 8;
	for(int i = 1; i <= n; i++)
		arr[i] = i;

	prepareTree(1, 1, n);

	for(int i = 1; i <= 15; i++) // total node = 15 after segmentation
		printf("%d ", tree[i]);
	int q = query(1, 1, n, 2, 4);

	printf("\nresult = %d", q);
	update(1, 1, n, 3, 3,55);
	printf("\n");
	
	for (int i =1; i <= 15; i++)
		printf("%d ", tree[i]);

	return 0;
}
